src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const copy = { ...currentState };
+        action.keysToRemove.forEach((key) => {
+          delete copy[key];
+        });
+
+        currentState = { ...copy };
+        break;
+
+      default:
+        break;
```

Comment:
The default case in the switch statement is empty. It's a good practice to handle unexpected `action.type` values. Consider throwing an error or logging a warning message to handle this case properly.